source/lac/petsc_vector_base.cc

```diff
@@ -125,11 +124,7 @@ namespace PETScWrappers
     : vector(nullptr)
     , ghosted(false)
     , last_action(::dealii::VectorOperation::unknown)
-  {
-    Assert(MultithreadInfo::is_running_single_threaded(),
-           ExcMessage("PETSc does not support multi-threaded access, set "
-                      "the thread limit to 1 in MPI_InitFinalize()."));
-  }
+  {}
```

Comment:
Have we addressed all places that didn't use to be thread-safe?

The origin of the claim that PETSc is not thread-safe is that there were places where we access individual vector elements by getting a pointer to the underlying PETSc array that stores the elements, set the element as desired, and then give the pointer back to PETSc. This was not thread-safe because PETSc temporarily gives up ownership of the pointer (and then reclaims it), and two threads can't do this get-pointer/access-element/return-pointer operation at the same time.

I don't know whether we still do that.

Reply:
As far as I understood the multi-threaded assembly used in this test, there are worker threads that compute element contributions that are then passed to a master thread that performs the needed `VecSetValues/MatSetValues` calls. This is supported in PETSc. What is not thread-safe is the access to PETSc internal data structures used for logging and other safety checks, which I have addressed here https://gitlab.com/petsc/petsc/-/merge_requests/5898

Reply:
I see. I believe that the facilities use in this specific test (the "WorkStream version 3 interface") does not use a central thread that does the sets/adds, but that individual threads all access elements independently, but that the work is split onto threads in such a way that two threads never access the same elements.

But it isn't just about this one test. We want to allow concurrent access to different elements of matrices and vectors. Is this now supported? I've looked into the history of the file, and the functions that were originally problematic were the ones that get and restore a "local form", like here:
```
    VectorReference::operator PetscScalar() const
    {
      AssertIndexRange(index, vector.size());

      // The vector may have ghost entries. In that case, we first need to
      // figure out which elements we own locally, then get a pointer to the
      // elements that are stored here (both the ones we own as well as the
      // ghost elements). In this array, the locally owned elements come first
      // followed by the ghost elements whose position we can get from an
      // index set.
      if (vector.ghosted)
        {
          PetscInt       begin, end;
          PetscErrorCode ierr =
            VecGetOwnershipRange(vector.vector, &begin, &end);
          AssertThrow(ierr == 0, ExcPETScError(ierr));

          Vec locally_stored_elements = PETSC_NULL;
          ierr = VecGhostGetLocalForm(vector.vector, &locally_stored_elements);        *****************
          AssertThrow(ierr == 0, ExcPETScError(ierr));

          PetscInt lsize;
          ierr = VecGetSize(locally_stored_elements, &lsize);
          AssertThrow(ierr == 0, ExcPETScError(ierr));

          PetscScalar *ptr;
          ierr = VecGetArray(locally_stored_elements, &ptr);           *********************
          AssertThrow(ierr == 0, ExcPETScError(ierr));

          PetscScalar value;

          if (index >= static_cast<size_type>(begin) &&
              index < static_cast<size_type>(end))
            {
              // local entry
              value = *(ptr + index - begin);
            }
          else
            {
              // ghost entry
              Assert(vector.ghost_indices.is_element(index),
                     ExcMessage(
                       "You are trying to access an element of a vector "
                       "that is neither a locally owned element nor a "
                       "ghost element of the vector."));
              const size_type ghostidx =
                vector.ghost_indices.index_within_set(index);

              AssertIndexRange(ghostidx + end - begin, lsize);
              value = *(ptr + ghostidx + end - begin);
            }

          ierr = VecRestoreArray(locally_stored_elements, &ptr);         ****************************
          AssertThrow(ierr == 0, ExcPETScError(ierr));

          ierr =
            VecGhostRestoreLocalForm(vector.vector, &locally_stored_elements);   *********************
          AssertThrow(ierr == 0, ExcPETScError(ierr));

          return value;
        }
  ...
```
My understanding -- possibly wrong -- has always been that these `GetLocalForm` and `GetArray` functions should not be called concurrently on different threads, but that the `Get/Restore` functions in essence delineate a critical section that in a multithreaded context need to be protected by a mutex. We don't have the mutex right now, and so my belief was that we don't support multiple threads.